### ISSUE TITLE
CacheV2: Unable to find pipeline caching scopes.

### ISSUE THREAD

**Issue by Gerrit-K** [microsoft/azure-pipelines-tasks#13420]
## Required Information

**Question, Bug, or Feature?**  
*Type*: Bug (essentially a reopen of #11869)

**Enter Task Name**: CacheV2

## Environment
- Server - Azure Pipelines
- Agent - Hosted
    - ubuntu-latest and Ubuntu16, agent version 2.173.0

## Issue Description
Same as #11869: when using the Cache task, I get the error "Unable to find pipeline caching scopes." (see logs below).

The pipeline is a pure YAML multi-stage pipeline. The problematic stage looks like this:

```yaml
stages:
  - stage: deploy_app
    displayName: Deploy app
    condition: and(succeeded(), or(eq(variables['Build.SourceBranch'], 'refs/heads/master'), eq(variables['DEPLOY'], 'true')))
    jobs:
      - deployment:
        displayName: Deploy
        environment: ${{ parameters.environment }}
        variables:
          - template: /infrastructure/azure-pipelines/state-secrets-variables.yaml
        strategy:
          runOnce:
            deploy:
              steps:
                - checkout: self
                - download: current
                  artifact: drop
[...]
                - task: Bash@3
                  displayName: Set npm_config_cache
                  inputs:
                    targetType: 'inline'
                    script: |
                      echo "##vso[task.setvariable variable=npm_config_cache]$(Pipeline.Workspace)/.npm"
                - task: Cache@2
                  displayName: Restore npm cache
                  inputs:
                    key: 'v2 | npm | "$(Agent.OS)" | infrastructure/package-lock.json'
                    path: '$(npm_config_cache)'
                    cacheHitVar: 'PULUMI_NODE_CACHE_RESTORED'
                  condition: and(succeeded(), ne(variables.PULUMI_NODE_CACHE_RESTORED, 'true'))
[...]
```

This issue first appeared after switching the `job` to a `deployment` job and adding an `environment`. 

While the Cache task is nested in a template multiple levels below, it does not fail on all pipelines that use the template. The directory structure is roughly as follows:
* /
  * ts-apps/
    * [...]/pipeline.yaml (these all work fine)
  * python-apps/
    * [...]/pipeline.yaml (these don't)
  * infrastructure/
    * azure-pipelines/
      * [...]
      * pulumi-command.yaml (this is the template that contains the Cache task and it's being used in all other pipelines)

### Task logs

This is a log of the task in one of the failing pipelines:

```
2020-08-12T11:24:32.8762742Z ##[debug]Evaluating condition for step: 'Restore npm cache'
2020-08-12T11:24:32.8768744Z ##[debug]Evaluating: and(succeeded(), ne(variables['PULUMI_NODE_CACHE_RESTORED'], 'true'))
2020-08-12T11:24:32.8769321Z ##[debug]Evaluating and:
2020-08-12T11:24:32.8777713Z ##[debug]..Evaluating succeeded:
2020-08-12T11:24:32.8778640Z ##[debug]..=> True
2020-08-12T11:24:32.8779079Z ##[debug]..Evaluating ne:
2020-08-12T11:24:32.8781238Z ##[debug]....Evaluating indexer:
2020-08-12T11:24:32.8807853Z ##[debug]......Evaluating variables:
2020-08-12T11:24:32.8811680Z ##[debug]......=> Object
2020-08-12T11:24:32.8812949Z ##[debug]......Evaluating String:
2020-08-12T11:24:32.8814043Z ##[debug]......=> 'PULUMI_NODE_CACHE_RESTORED'
2020-08-12T11:24:32.8818618Z ##[debug]....=> Null
2020-08-12T11:24:32.8819553Z ##[debug]....Evaluating String:
2020-08-12T11:24:32.8820025Z ##[debug]....=> 'true'
2020-08-12T11:24:32.8826486Z ##[debug]....=> Unable to coerce String to Null.
2020-08-12T11:24:32.8826957Z ##[debug]..=> True
2020-08-12T11:24:32.8827774Z ##[debug]=> True
2020-08-12T11:24:32.8830742Z ##[debug]Expanded: and(True, ne(Null, 'true'))
2020-08-12T11:24:32.8831265Z ##[debug]Result: True
2020-08-12T11:24:32.8831873Z ##[section]Starting: Restore npm cache
2020-08-12T11:24:32.8837212Z ==============================================================================
2020-08-12T11:24:32.8837606Z Task         : Cache
2020-08-12T11:24:32.8837881Z Description  : Cache files between runs
2020-08-12T11:24:32.8838181Z Version      : 2.0.1
2020-08-12T11:24:32.8838452Z Author       : Microsoft Corporation
2020-08-12T11:24:32.8838819Z Help         : https://aka.ms/pipeline-caching-docs
2020-08-12T11:24:32.8839472Z ==============================================================================
2020-08-12T11:24:33.2287914Z Resolving key:
2020-08-12T11:24:33.2443034Z  - v2                                      [string]
2020-08-12T11:24:33.2443678Z  - npm                                     [string]
2020-08-12T11:24:33.2446565Z  - "Linux"                                 [string]
2020-08-12T11:24:33.2934066Z  - infrastructure/package-lock.json [file] --> C416E6FA791258733E1AC45409FCD865031CD54CBB0DBA3E05D52F998149B3A2
2020-08-12T11:24:33.2967388Z Resolved to: v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=
2020-08-12T11:24:33.2984945Z ##[debug]Processed: ##vso[task.settaskvariable variable=RESTORE_STEP_RAN;issecret=False;]true
2020-08-12T11:24:33.2986084Z ##[debug]Processed: ##vso[task.settaskvariable variable=RESTORE_STEP_RESOLVED_FINGERPRINT;issecret=False;]v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=
2020-08-12T11:24:33.5086521Z Verbose, ArtifactHttpClientFactory.CreateVssHttpClient: DedupStoreHttpClient with BaseUri: https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/, MaxRetries:5, SendTimeout:00:00:50
2020-08-12T11:24:33.5107103Z ##[debug]Dedup parallelism: 192
2020-08-12T11:24:33.6446910Z Information, ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session 30e987a5-fdc0-495b-a563-3e9bdfa55eda
2020-08-12T11:24:33.6447426Z Verbose, Started ITelemetrySender BlobStoreApplicationInsightsTelemetrySender.
2020-08-12T11:24:34.1745826Z Information, Getting a pipeline cache artifact with one of the following fingerprints:
2020-08-12T11:24:34.1746828Z Information, Fingerprint: `v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=`
2020-08-12T11:24:34.3138190Z Verbose, ApplicationInsightsTelemetrySender waiting for 0 of 2 TrackEvent operations to complete...
2020-08-12T11:24:35.3672290Z Verbose, ApplicationInsightsTelemetrySender sent PipelineCache.RestoreCache telemetry
2020-08-12T11:24:35.9352017Z Verbose, ApplicationInsightsTelemetrySender sent Error telemetry
2020-08-12T11:24:35.9359827Z Verbose, ApplicationInsightsTelemetrySender operations completed in 1621 ms
2020-08-12T11:24:35.9363777Z Information, ApplicationInsightsTelemetrySender correlated 2 events with X-TFS-Session 30e987a5-fdc0-495b-a563-3e9bdfa55eda
2020-08-12T11:24:35.9364959Z Verbose, Stopped ITelemetrySender BlobStoreApplicationInsightsTelemetrySender.
2020-08-12T11:24:35.9414318Z ##[error]Unable to find pipeline caching scopes.
2020-08-12T11:24:35.9429197Z ##[debug]Processed: ##vso[task.logissue type=error;]Unable to find pipeline caching scopes.
2020-08-12T11:24:35.9431984Z ##[debug]Processed: ##vso[task.complete result=Failed;]
2020-08-12T11:24:35.9436790Z ##[debug]   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.HandleResponseAsync(HttpResponseMessage response, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync(HttpRequestMessage message, HttpCompletionOption completionOption, Object userState, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpRequestMessage message, Object userState, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.WebApi.VssHttpClientBase.SendAsync[T](HttpMethod method, IEnumerable`1 additionalHeaders, Guid locationId, Object routeValues, ApiResourceVersion version, HttpContent content, IEnumerable`1 queryParameters, Object userState, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.PipelineCache.WebApi.PipelineCacheClient.GetPipelineCacheArtifactAsync(Fingerprint[] fingerprints, CancellationToken cancellationToken)
   at Microsoft.VisualStudio.Services.PipelineCache.WebApi.PipelineCacheClient.<>c__DisplayClass9_0.<<GetPipelineCacheArtifactAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Services.BlobStore.Common.Telemetry.BlobStoreClientTelemetry.MeasureActionAsync[TResult](BlobStoreTelemetryRecord record, Func`1 actionAsync)
   at Microsoft.VisualStudio.Services.Content.Common.ExceptionExtensions.ReThrow(Exception ex)
   at Microsoft.VisualStudio.Services.BlobStore.Common.Telemetry.BlobStoreClientTelemetry.MeasureActionAsync[TResult](BlobStoreTelemetryRecord record, Func`1 actionAsync)
   at Microsoft.VisualStudio.Services.PipelineCache.WebApi.PipelineCacheClient.GetPipelineCacheArtifactAsync(Fingerprint[] fingerprints, CancellationToken cancellationToken, PipelineCacheTelemetryRecord cacheRecordOptional)
   at Agent.Plugins.PipelineCache.PipelineCacheServer.DownloadAsync(AgentTaskPluginExecutionContext context, Fingerprint[] fingerprints, String path, String cacheHitVariable, CancellationToken cancellationToken)
   at Agent.Plugins.PipelineCache.RestorePipelineCacheV0.ProcessCommandInternalAsync(AgentTaskPluginExecutionContext context, Fingerprint fingerprint, Func`1 restoreKeysGenerator, String path, CancellationToken token)
   at Agent.Plugins.PipelineCache.PipelineCacheTaskPluginBase.RunAsync(AgentTaskPluginExecutionContext context, CancellationToken token)
   at Agent.PluginHost.Program.Main(String[] args)
2020-08-12T11:24:35.9474315Z ##[section]Finishing: Restore npm cache
```

And this is the log of the task in one of the working pipelines (note the identical cache fingerprints):

```
2020-08-12T11:42:00.6311714Z ##[debug]Evaluating condition for step: 'Restore npm cache'
2020-08-12T11:42:00.6316188Z ##[debug]Evaluating: and(succeeded(), ne(variables['PULUMI_NODE_CACHE_RESTORED'], 'true'))
2020-08-12T11:42:00.6316820Z ##[debug]Evaluating and:
2020-08-12T11:42:00.6323659Z ##[debug]..Evaluating succeeded:
2020-08-12T11:42:00.6324156Z ##[debug]..=> True
2020-08-12T11:42:00.6324487Z ##[debug]..Evaluating ne:
2020-08-12T11:42:00.6325848Z ##[debug]....Evaluating indexer:
2020-08-12T11:42:00.6338979Z ##[debug]......Evaluating variables:
2020-08-12T11:42:00.6341452Z ##[debug]......=> Object
2020-08-12T11:42:00.6342292Z ##[debug]......Evaluating String:
2020-08-12T11:42:00.6342922Z ##[debug]......=> 'PULUMI_NODE_CACHE_RESTORED'
2020-08-12T11:42:00.6345795Z ##[debug]....=> Null
2020-08-12T11:42:00.6346307Z ##[debug]....Evaluating String:
2020-08-12T11:42:00.6346611Z ##[debug]....=> 'true'
2020-08-12T11:42:00.6350673Z ##[debug]....=> Unable to coerce String to Null.
2020-08-12T11:42:00.6351026Z ##[debug]..=> True
2020-08-12T11:42:00.6351536Z ##[debug]=> True
2020-08-12T11:42:00.6353510Z ##[debug]Expanded: and(True, ne(Null, 'true'))
2020-08-12T11:42:00.6353930Z ##[debug]Result: True
2020-08-12T11:42:00.6354314Z ##[section]Starting: Restore npm cache
2020-08-12T11:42:00.6358728Z ==============================================================================
2020-08-12T11:42:00.6359059Z Task         : Cache
2020-08-12T11:42:00.6359315Z Description  : Cache files between runs
2020-08-12T11:42:00.6359559Z Version      : 2.0.1
2020-08-12T11:42:00.6359808Z Author       : Microsoft Corporation
2020-08-12T11:42:00.6360208Z Help         : https://aka.ms/pipeline-caching-docs
2020-08-12T11:42:00.6360534Z ==============================================================================
2020-08-12T11:42:00.9194575Z Resolving key:
2020-08-12T11:42:00.9333444Z  - v2                                      [string]
2020-08-12T11:42:00.9334194Z  - npm                                     [string]
2020-08-12T11:42:00.9336407Z  - "Linux"                                 [string]
2020-08-12T11:42:00.9753932Z  - infrastructure/package-lock.json [file] --> C416E6FA791258733E1AC45409FCD865031CD54CBB0DBA3E05D52F998149B3A2
2020-08-12T11:42:00.9783475Z Resolved to: v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=
2020-08-12T11:42:00.9797253Z ##[debug]Processed: ##vso[task.settaskvariable variable=RESTORE_STEP_RAN;issecret=False;]true
2020-08-12T11:42:00.9798621Z ##[debug]Processed: ##vso[task.settaskvariable variable=RESTORE_STEP_RESOLVED_FINGERPRINT;issecret=False;]v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=
2020-08-12T11:42:01.1781274Z Verbose, ArtifactHttpClientFactory.CreateVssHttpClient: DedupStoreHttpClient with BaseUri: https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/, MaxRetries:5, SendTimeout:00:00:50
2020-08-12T11:42:01.1818567Z ##[debug]Dedup parallelism: 192
2020-08-12T11:42:01.3113826Z Information, ApplicationInsightsTelemetrySender will correlate events with X-TFS-Session a86db722-8664-45cc-9092-3b6564825b60
2020-08-12T11:42:01.3114834Z Verbose, Started ITelemetrySender BlobStoreApplicationInsightsTelemetrySender.
2020-08-12T11:42:01.6837124Z Information, Getting a pipeline cache artifact with one of the following fingerprints:
2020-08-12T11:42:01.6838105Z Information, Fingerprint: `v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=`
2020-08-12T11:42:02.1604953Z Information, There is a cache hit: `v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=`
2020-08-12T11:42:02.1607371Z Information, Used scope: 259;f5c7a962-6082-42ca-ac3f-137df246978a;refs/heads/master;78fa3cd1-67a1-4ad7-876c-fd5872ec6db5
2020-08-12T11:42:02.1609421Z Information, Missed on the following scopes: 259;f5c7a962-6082-42ca-ac3f-137df246978a;refs/heads/pipeline-environments;78fa3cd1-67a1-4ad7-876c-fd5872ec6db5
2020-08-12T11:42:02.2345897Z ##[debug]Processed: ##vso[telemetry.publish area=AzurePipelinesAgent;feature=PipelineCache]{"FileCount":"0","PlanId":"212baa79-b64f-4e48-a582-b9fbbe790e31","JobId":"7e4fa394-a526-5b11-21c4-3735ed6aef8d","TaskInstanceId":"50a0e04b-2bf7-53cf-5fc7-a2185c6eb453","CacheResult":"Hit","ActionDurationMs":"484","ActionName":"PipelineCache.RestoreCache","ActionResult":"Success","AttemptNumber":"1","ItemCount":"0","Level":"ThirdParty","CreatedUtcNow":"2020-08-12T11:42:01.6728423Z","SentUtcNow":"2020-08-12T11:42:02.1628399Z","BaseAddress":"https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/","X_TFS_Session":"a86db722-8664-45cc-9092-3b6564825b60","DeploymentEnvironment":"PRODUCTION","DeploymentEnvironmentIsProduction":"True","VSOAccount":"vsblobprodsu6weu","OSName":"Linux","OSVersion":"2020","FrameworkDescription":".NET Core ","ProcessName":"Agent.PluginHost","DotNetReleaseDword":"-1","Version":"18.169.30119.0","ExceptionCount":"0"}
2020-08-12T11:42:02.2348658Z Entry found at fingerprint: `v2|npm|"Linux"|F/PBNr3ki2VwlGEI/ltlXjIXpRWDf8v9fCXR3SHt4fg=`
2020-08-12T11:42:02.2349038Z Verbose, PipelineCache.DownloadAsync starting
2020-08-12T11:42:02.2952828Z Verbose, ArtifactHttpRetryMessageHandler.TraceHttpRequestFailed: https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/_apis/dedup/chunks/9524D78CD73AA02ADA29A9AC5B04E255BE474724688AA1D44687AB0C11FAFE2A01?redirect=true, HttpStatusCode.RedirectMethod, activity 
2020-08-12T11:42:02.4496173Z ##[debug]Starting 'tar' with arguments '-v -xf - -C .'...
2020-08-12T11:42:02.4854943Z Verbose, ArtifactHttpRetryMessageHandler.TraceHttpRequestFailed: https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/_apis/dedup/nodes/45ED9CE75A7D1DE2FDD2A5E949FADA22C44122EB7E8385EE859B5E94A7D1783E02?redirect=true, HttpStatusCode.RedirectMethod, activity 
2020-08-12T11:42:02.5828232Z Information, Expected size to be downloaded: 19.6 MB
2020-08-12T11:42:02.6364412Z Information, Downloaded 0.0 MB out of 19.6 MB (0%).
2020-08-12T11:42:02.9638572Z Verbose, ApplicationInsightsTelemetrySender sent PipelineCache.RestoreCache telemetry
2020-08-12T11:42:03.9737568Z ./
2020-08-12T11:42:03.9738297Z ./_cacache/
[...]
2020-08-12T11:42:04.5091120Z Information, Downloaded 19.6 MB out of 19.6 MB (100%).
2020-08-12T11:42:04.5091362Z Information, 
2020-08-12T11:42:04.5091573Z Download statistics:
2020-08-12T11:42:04.5091806Z Total Content: 19.6 MB
2020-08-12T11:42:04.5092053Z Physical Content Downloaded: 19.1 MB
2020-08-12T11:42:04.5092324Z Compression Saved: 0.5 MB
2020-08-12T11:42:04.5092561Z Local Caching Saved: 0.0 MB
2020-08-12T11:42:04.5092809Z Chunks Downloaded: 250
2020-08-12T11:42:04.5093023Z Nodes Downloaded: 1
[...]
2020-08-12T11:42:04.5122485Z Process exit code: 0
2020-08-12T11:42:04.5122785Z Verbose, PipelineCache.DownloadAsync completed in 2.3077915 seconds
2020-08-12T11:42:04.5535807Z ##[debug]Processed: ##vso[telemetry.publish area=AzurePipelinesAgent;feature=PipelineCache]{"FileCount":"0","PlanId":"212baa79-b64f-4e48-a582-b9fbbe790e31","JobId":"7e4fa394-a526-5b11-21c4-3735ed6aef8d","TaskInstanceId":"50a0e04b-2bf7-53cf-5fc7-a2185c6eb453","ActionDurationMs":"2307","ActionName":"PipelineCache.DownloadAsync","ActionResult":"Success","AttemptNumber":"1","ItemCount":"0","Level":"ThirdParty","CreatedUtcNow":"2020-08-12T11:42:02.1692955Z","SentUtcNow":"2020-08-12T11:42:04.4802795Z","BaseAddress":"https://vsblobprodsu6weu.vsblob.visualstudio.com/Ad7b6fe08-347e-4cca-91fb-d5c313d16196/","X_TFS_Session":"a86db722-8664-45cc-9092-3b6564825b60","DeploymentEnvironment":"PRODUCTION","DeploymentEnvironmentIsProduction":"True","VSOAccount":"vsblobprodsu6weu","OSName":"Linux","OSVersion":"2020","FrameworkDescription":".NET Core ","ProcessName":"Agent.PluginHost","DotNetReleaseDword":"-1","Version":"18.169.30119.0","ExceptionCount":"0"}
2020-08-12T11:42:04.5538630Z Cache restored.
2020-08-12T11:42:04.5540163Z ##[debug]Processed: ##vso[task.setvariable variable=PULUMI_NODE_CACHE_RESTORED;issecret=False;]true
2020-08-12T11:42:04.5540788Z Verbose, ApplicationInsightsTelemetrySender waiting for 0 of 2 TrackEvent operations to complete...
2020-08-12T11:42:05.7496500Z Verbose, ApplicationInsightsTelemetrySender sent PipelineCache.DownloadAsync telemetry
2020-08-12T11:42:05.7496996Z Verbose, ApplicationInsightsTelemetrySender operations completed in 1260 ms
2020-08-12T11:42:05.7498152Z Information, ApplicationInsightsTelemetrySender correlated 2 events with X-TFS-Session a86db722-8664-45cc-9092-3b6564825b60
2020-08-12T11:42:05.7498578Z Verbose, Stopped ITelemetrySender BlobStoreApplicationInsightsTelemetrySender.
2020-08-12T11:42:05.7632583Z ##[section]Finishing: Restore npm cache
```

**Comment by johnterickson** [comment]
`Pipeline Caching` uses cache namespaces scoped to specific branches - details here: https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#cache-isolation-and-security.

@vtbassmatt to check me on this, but a `deployment` is tied to an artifact - not a source code branch. This is why no scopes are found.

Generally speaking, the flow is that the build stages create artifacts and the deployment stage deploys the given artifact.  To make the deployment repeatable (not that I need to describe this to a Pulumi user 😁), you would want the deployment to be self-contained by the artifact(s) going into it.  Is what you are restoring from the cache something that the build could be producing?

Other options include: Universal Packages (the same tech in a different wrapper) and https://github.com/Microsoft/azure-pipelines-artifact-caching-tasks .  Think very carefully about security (which identities/builds have access to write to the "cache") if you do so.

**Comment by Gerrit-K** [comment]
Thanks for the swift response!

I'm still a bit puzzled about your notes on branches, because 1) all of the questioned pipelines belong to one monorepo (=same branch) and 2) some of them are working while others aren't. If the cache namespaces are scoped to branches and `deployment` jobs are not tied to them, I would've expected that either _all_ or _none_ fail 🤔  especially because they all use the exact same task template. And the fact that _exactly_ the python pipelines fail whereas the typescript pipelines succeed seems _very_ suspicious to me and I bet I'm still overlooking something there.

Your note on the flow however brings up a valid point. Let me quickly outline our pipeline structure:
* Each pipeline has 3 stages: `check/test`, `build` and `deploy`
* The `build` stage creates an artifact for the `deploy` stage to consume
* The `deploy` stage however needs the Pulumi code/project in order to know where to deploy the artifact to. This is why we checkout the source branch even in this stage.

Since the Pulumi project is built in typescript, I thought it would be good practice to cache the npm cache (as described on the docs page you've linked), which is where I'm running into this issue. Technically, I could probably simply skip caching because the speedup isn't that notable, but now I'm curious why this isn't working.

(BTW: The reason for us to switch to `deployment` jobs was to use environments with exclusive locks, because the interference of multiple parallel jobs doing deployments caused havoc ...) 

Following your thoughts on artifacts, would it be better to use the build stage to pack the Pulumi project (with installed packages) into an artifact as well and provide it to the `deployment` job instead of checking out the sources?

**Comment by vtbassmatt** [comment]
(I believe @johnterickson is right about `deployment` not being tied to a triggering branch; @ashokirla or @shashban might be able to say more about how that context flows around in a unified pipeline where the `trigger` was a branch but one of the jobs is a `deployment`.)

**Comment by mmuffins** [comment]
We experience the same issue with the cache task in deployment jobs. Is there anything we can do about it (other than not using caching)?

**Comment by avishnyakov** [comment]
Hitting the same problem. We tend to use deployment due environments, so we can have traceability and other checks. 
Looks that Cache task is not supported when using deployments (eg, "release" pipelines) in YAML.

@vtbassmatt, team, what would be a path forward here?

**Comment by avishnyakov** [comment]
@johnterickson 

> Generally speaking, the flow is that the build stages create artifacts and the deployment stage deploys the given artifact. To make the deployment repeatable (not that I need to describe this to a Pulumi user 😁), you would want the deployment to be self-contained by the artifact(s) going into it. Is what you are restoring from the cache something that the build could be producing?

Split between build stages and deployment stage is reasonable, of course. Say, build stages produce some kind of configuration, an application, a docker image or any of that kind, an artifact which then deployed by the deployment stages. People might want to use docker images in the deployment stages to deploy artifacts coming from the build stages. Having same-same images which perform deployments is not a bad idea - pinning tools, having scripts and other helpers run under controlled, repeatable docker image. This is exactly the case where caching is very helpful at the release stages.

**Comment by cforce** [comment]
same problem here
Its unreliable. It work for the same tempalte for one usage and for another stage using the same tpl it doesn't ..

**Comment by GeorgeNarcis** [comment]
We are facing the same issue when using CacheV2 task inside a deployment job in multi stage pipelines. According to the docs "Caching is currently supported in CI and deployment jobs, but not classic release jobs."
So is this a bug or it doesn't work in deployment jobs ?

**Comment by johnterickson** [comment]
For security, Pipeline Caching relies on certain claims related to the branch name to be in the token.  The Pipelines service definitely puts these claims in the token when it runs build jobs.  I am unsure for which circumstances releases will have these token.  Maybe @vijayma knows. It's unlikely that releases that do not consume a `git` repo as an artifact would have the needed claims.

**Comment by dmcweeney** [comment]
Same problem here...  would love a reliable solution!

**Comment by GeorgeNarcis** [comment]
@johnterickson do you have any updates on this ? We have a lot of multi stage pipelines that use Powershell modules to do various tasks in the deployment steps, so we could really benefit from this if it would worked as stated in docs.

**Comment by JaydenMaalouf** [comment]
I know this is closed but looking for an update on this too.
Seeing this occur whenever we use Cache@2 in Deployment Jobs.

**Comment by cforce** [comment]
never give up :|

**Comment by flaviostutz** [comment]
Are we really abandoned in this one?

**Comment by bstri** [comment]
Ran into this issue personally and did some debugging to try to learn more about it. Came up with a minimal reproducible example. 

In it there is an optional build stage (specified by the `buildStage` parameter) that creates a folder and caches it under the key `cache`. Then there is a deploy stage that creates a different folder and caches it under the key `cache2`. 

The caching scopes error is only present when the `buildStage` parameter is set to `False`.
Note that the cache keys are different between build and deploy. So, your theory was pretty close @aholthagerty.

```yaml
parameters:
- name: buildStage
  type: boolean
  default: True

stages:

- ${{ if eq(parameters.buildStage, True) }}:
  - stage: build
    jobs:
    - job:
      steps:
      - task: Bash@3
        inputs:
          targetType: inline
          script: |
            mkdir -p $(Agent.TempDirectory)/cache

      - task: Cache@2
        inputs:
          key: cache | $(Build.SourceVersion)
          path: $(Agent.TempDirectory)/cache
          restoreKeys: |
            cache

- stage: deploy
  ${{ if eq(parameters.buildStage, True) }}:
    dependsOn: build
  jobs:
  - deployment: deploy
    environment: <environment_name_here>
    strategy:
      runOnce:
        deploy:
          steps:
          - checkout: self

          - task: Bash@3
            inputs:
              targetType: inline
              script: |
                mkdir -p $(Agent.TempDirectory)/cache2

          - task: Cache@2
            inputs:
              key: cache2 | $(Build.SourceVersion)
              path: $(Agent.TempDirectory)/cache2
              restoreKeys: |
                cache2
```

Another interesting piece of info: originally I did not use a buildStage parameter, and rather simply deselected the build stage when running my pipeline. Even though the build stage was skipped in that scenario, the cache step in the deploy stage did NOT fail.

**Comment by manne** [comment]
We have the same issue here. It would be awesome that caching works as expected

**Comment by jadhavnitind** [comment]
Same here. Any solution on this ?

**Comment by VazquezAgustin** [comment]
> I have a theory that the cache task does not work in a deployment job _unless_ there is a cache task in an earlier non-deployment job using the same `key` within the pipeline.

Same issue and same theory. When running my build (non deployment job), the caching works. If the pipeline only uses deployment jobs it fails.

**Comment by VazquezAgustin** [comment]
Just figured out that the non deployment job can be empty, that could be a workaround that you can use

**Comment by tcinbis** [comment]
@johnterickson sorry for the ping, we are also experiencing this issue and I want add some context why it would make sense imho to have the caching task work out of the box in deployment jobs as well.

We are using Pulumiwith go to deploy our infrastructure. Hence the use of deployment task. Pulumi will fetch the go dependencies during deployment.

**Comment by abij** [comment]
As a work around, I'm using this. It's working in a template for jobs _and_ deployments.

```yaml
- task: Cache@2
  displayName: "Cache load"
  condition: eq(variables['Environment.Name'], '')  # Job types deployment provide this variable, deployments cannot cache.
  inputs:
    cacheHitVar: cacheRestored    
    key: '<useful key including "$(Agent.OS)" and a version nr>' 
    path: $(Pipeline.Workspace)/.cache_folder

- bash: |
    ...  # download stuff in $(Pipeline.Workspace)/.cache_folder
  displayName: "Not cached"
  condition: not(eq(variables.cacheRestored, 'true'))   # When deployment job, value is empty and therefor is not skipped.
```

**Comment by Green0wl** [comment]
the issue is still there.

**Comment by Riaku** [comment]
Still there in 2025 :( seems to be related to deployment jobs

**Comment by chris-pinola-rf** [comment]
The odds of getting any traction on this closed ticket feel slim to me, so I've reopened it and will do my best to keep it from getting closed as stale: #21567
